sympy/simplify/cse_main.py

```diff
@@ -136,6 +137,238 @@ def postprocess_for_cse(expr, optimizations):
     return expr
 
 
+class FuncArgTracker(object):
+    """
+    A class which manages a mapping from functions to arguments and an inverse
+    mapping from arguments to functions.
+    """
+
+    def __init__(self, funcs):
+        # To minimize the number of symbolic comparisons, all function arguments
+        # get assigned a value number.
+        self.value_numbers = {}
+        self.value_number_to_value = []
+
+        # Both of these maps use integer indices for arguments / functions.
+        self.arg_to_funcset = []
+        self.func_to_argset = []
+
+        for func_i, func in enumerate(funcs):
+            func_argset = OrderedSet()
+
+            for func_arg in func.args:
+                arg_number = self.get_or_add_value_number(func_arg)
+                func_argset.add(arg_number)
+                self.arg_to_funcset[arg_number].add(func_i)
+
+            self.func_to_argset.append(func_argset)
+
+    def get_args_in_value_order(self, argset):
+        """
+        Return the list of arguments in sorted order according to their value
+        numbers.
+        """
+        return [self.value_number_to_value[argn] for argn in sorted(argset)]
+
+    def get_or_add_value_number(self, value):
```

Comment:
rather than introducing a new name, can we just use the already known `setdefault` as is used for dictionaries?

```
>>> d = {1: 2}
>>> d.setdefault(1, 0)
2
>>> d.setdefault(3, 0)
0
>>> d
{1: 2, 3: 0}
```

Reply:
This would introduce variation wrt sumpy...which might get updated and then it might be harder to update this. So I'm OK with it as it is. Are you ready to commit this?

Reply:
Yes, it'll be easier if the implementations were same.